Fix SegmentReplicationUsingRemoteStoreIT#testDropPrimaryDuringReplication.

mch2 opened the pull request:
### Description
This test is yet again failing but only with remote store.  This time a concurrent flush can wipe out an old commit file while we are in the remote store refresh listener. The listener will fetch the latest infos from the reader which will reference a segments_n that has been deleted by an incoming flush.

To fix this, InternalEngine will conditionally acquire the previous commit point and preserve it until a new commit is loaded onto the reader.  This guarantees a commit point is not deleted while inside of an active refresh.

### Related Issues
Resolves #https://github.com/opensearch-project/OpenSearch/issues/8059

### Check List
- [x] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

mch2 commented:
pushed a rebase here - tagging @sachinpkale @ashking94 for review on this.

sachinpkale commented on server/src/main/java/org/opensearch/index/engine/InternalEngine.java:
```diff
@@ -2156,15 +2148,26 @@ protected SegmentInfos getLatestSegmentInfos() {
+            SegmentInfos readerInfos = ((StandardDirectoryReader) reader.getDelegate()).getSegmentInfos();
+            // if the latest index commit on disk has a higher gen than the reader load that instead.
+            // This ensures our commit file will exist throughout upload process.
+            if (lastIndexCommit.get().getGeneration() > readerInfos.getGeneration()) {
```
In what scenarios this `if` condition will be true? Flush created new commit but reader is not yet updated?

I am trying to understand if we discard newly created segment files by discarding `readerInfos`. This can result in data loss. Is it possible to add an assert to validate this?

mch2 replied:
we can assert the SI versions are the same or that the commit is ahead.

This happens when:
1. index docs & refresh (R1) flushing new segments to disk and updating the reader.
2. During refresh listeners for R1 a new commit comes and wipes the segments_n on disk, replacing with segments_n+1
3. Still in refresh listeners we acquire a snapshot.  The reader still references segments_n while it is not on disk. (nosuchfile thrown in remote store listener).
4. Refresh R1 would complete - then Refresh R2 (blockign refresh after flush) will execute updating the reader to the latest infos but only if there are segment updates - didRefresh will return false here if this is an empty commit because nrt current check only considers the SI version not gen.


An alternative we could try here is to acquire the latest commit ref before refresh occurs in the refresh listener, and release it after refresh completes.  However, I think we could still at any point have a segments_n ref'd on the reader that is not on disk given flush can occur at any time.

ashking94 replied:
on point 4, if there are no segment updates, on a flush will it trigger a refresh? I did not follow the point completely. But are we saying that if R2 occurs on account of a flush (with no segment changes), then the readers info will not get updated and will continue to remain stale, while the latest commit has gone to next generation?

ashking94 commented on server/src/main/java/org/opensearch/index/engine/InternalEngine.java:
```diff
@@ -2156,15 +2148,26 @@ protected SegmentInfos getLatestSegmentInfos() {
     @Override
     public GatedCloseable<SegmentInfos> getSegmentInfosSnapshot() {
-        final SegmentInfos segmentInfos = getLatestSegmentInfos();
+        final SegmentInfos segmentInfos;
```
lets update the java doc for this method. The dependency order of this method and  `getLatestSegmentInfos` has reversed.

ashking94 commented on server/src/main/java/org/opensearch/index/engine/InternalEngine.java:
```diff
@@ -2156,15 +2148,26 @@ protected SegmentInfos getLatestSegmentInfos() {
+                segmentInfos = readerInfos;
+            }
         } catch (IOException e) {
             throw new EngineException(shardId, e.getMessage(), e);
```
Should we also release the reader in case if there are any exception b/w line number 2156 and 2164

mch2 replied:
good points, will update here

ashking94 commented on server/src/main/java/org/opensearch/index/engine/InternalEngine.java:
```diff
@@ -2156,15 +2148,26 @@ protected SegmentInfos getLatestSegmentInfos() {
         return new GatedCloseable<>(segmentInfos, () -> {
             try {
-                indexWriter.decRefDeleter(segmentInfos);
+                internalReaderManager.release(reader);
```
If we are relying on last index commit to provide the segment infos (when if condition holds true), should we call the `internalReaderManager.release(reader)` earlier itself? Or are there any implications of releasing vs not releasing?

mch2 replied:
good callout as well, we don't need the reader ref as the segments will be preserved by the commit.

ashking94 commented on server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java:
```diff
@@ -211,13 +211,14 @@ private boolean syncSegments() {
-                        + segmentInfos.getGeneration()
-                        + " does not match metadata generation: "
-                        + checkpoint.getSegmentsGen();
+                    assert segmentInfos.getVersion() == checkpoint.getSegmentInfosVersion() : "SegmentInfos version: "
+                        + segmentInfos.getVersion()
+                        + " does not match metadata version: "
+                        + checkpoint.getSegmentInfosVersion();
```
The version inside the ReplicationCheckpoint object seem to be created from the same segment infos that is being returned from the `getLatestSegmentInfosAndCheckpoint` method. Looks like the assertion evaluates to true (always?).

mch2 replied:
you're right - this is leftover from when we were computing the checkpoint earlier.  Would like to take https://github.com/opensearch-project/OpenSearch/issues/9046 so we can simplify here.

ashking94 commented on server/src/main/java/org/opensearch/indices/replication/common/CopyState.java:
```diff
@@ -65,10 +62,6 @@ public CopyState(ReplicationCheckpoint requestedReplicationCheckpoint, IndexShar
     protected void closeInternal() {
         try {
             segmentInfosRef.close();
-            // commitRef may be null if there were no pending delete files
```
Why are we removing the below code?

mch2 replied:
The commit ref here is redundant because its now held inside of IE when getting the snapshot.

I think another refactor opportunity here is to also use this CopyState for remote store, and build it once after IE refreshes and store a ref to it.  This would eliminate the need for fetching & computing SI snapshots outside of refresh listeners.

ashking94 reviewed (commented):
Have left some question, yet to review tests and ITs.

mch2 commented on server/src/main/java/org/opensearch/index/engine/InternalEngine.java:
```diff
@@ -2156,15 +2148,26 @@ protected SegmentInfos getLatestSegmentInfos() {
-        final SegmentInfos segmentInfos = getLatestSegmentInfos();
+        final SegmentInfos segmentInfos;
+        final OpenSearchDirectoryReader reader;
+        final GatedCloseable<IndexCommit> lastIndexCommit = acquireLastIndexCommit(false);
```
yes it is possible, but acquiring here will ensure we are not deleting the previous gen on a new commit.

mch2 commented:
@ashking94 @sachinpkale I've changed this to be much more simple and acquire the latest commit before an IE makes another and hold it until the new commit is refreshed on.  This ensures that any segments_n fetched with getSegmentInfosSnapshot is guaranteed to exist until the refresh completes.

Also apologies for the force push here, I updated the commit message on the original commit to reflect this change.

mch2 commented:
Recent GC failures are all https://github.com/opensearch-project/OpenSearch/issues/9407

mch2 commented:
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** UNSTABLE ❕
> * **TEST FAILURES:**
> 
> ```
>       1 org.opensearch.client.PitIT.testDeleteAllAndListAllPits
>       1 org.opensearch.action.admin.cluster.node.tasks.ResourceAwareTasksTests.testTaskResourceTrackingDuringTaskCancellation
> ```
> 
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/23743/
> * **CommitID:** [67d0701](https://github.com/opensearch-project/OpenSearch/commit/67d0701ea16043fff5f0fbbeda1859eb35c9c85e)
>   Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

https://github.com/opensearch-project/OpenSearch/issues/5030
and
https://github.com/opensearch-project/OpenSearch/issues/5329
